homeassistant/components/ws66i/media_player.py

```diff
@@ -0,0 +1,199 @@
+"""Support for interfacing with WS66i 6 zone home audio controller."""
+from datetime import timedelta
+import logging
+
+from homeassistant import core
+from homeassistant.components.media_player import MediaPlayerEntity
+from homeassistant.components.media_player.const import (
+    SUPPORT_SELECT_SOURCE,
+    SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON,
+    SUPPORT_VOLUME_MUTE,
+    SUPPORT_VOLUME_SET,
+    SUPPORT_VOLUME_STEP,
+)
+from homeassistant.const import CONF_IP_ADDRESS, STATE_OFF, STATE_ON
+from homeassistant.helpers import entity_platform
+
+from .const import CONF_SOURCES, DOMAIN, SERVICE_RESTORE, SERVICE_SNAPSHOT, WS66I_OBJECT
+
+_LOGGER = logging.getLogger(__name__)
+
+PARALLEL_UPDATES = 1
+
+SUPPORT_WS66I = (
+    SUPPORT_VOLUME_MUTE
+    | SUPPORT_VOLUME_SET
+    | SUPPORT_VOLUME_STEP
+    | SUPPORT_TURN_ON
+    | SUPPORT_TURN_OFF
+    | SUPPORT_SELECT_SOURCE
+)
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+
+@core.callback
+def _get_sources_from_dict(data):
+    sources_config = data[CONF_SOURCES]
+
+    # Dict index to custom name
+    source_id_name = {int(index): name for index, name in sources_config.items()}
+
+    # Dict custom name to index
+    source_name_id = {v: k for k, v in source_id_name.items()}
+
+    # List of custom names
+    source_names = sorted(source_name_id.keys(), key=lambda v: source_name_id[v])
+
+    return [source_id_name, source_name_id, source_names]
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the WS66i 6-zone amplifier platform."""
+    ip_addr = config_entry.data[CONF_IP_ADDRESS]
+    ws66i = hass.data[DOMAIN][config_entry.entry_id][WS66I_OBJECT]
+    sources = _get_sources_from_dict(config_entry.options)
+
+    # Build the entities that control each of the zones.
+    # Zones 11 - 16 are the master amp
+    # Zones 21,31 - 26,36 are the daisy-chained amps
+    entities = []
+    for amp_num in range(1, 4):
+
+        if amp_num > 1:
+            # Don't add entities that aren't present
+            status = await hass.async_add_executor_job(
+                ws66i.zone_status, (amp_num * 10 + 1)
+            )
+            if status is None:
+                break
+
+        _LOGGER.info("Detected amp %d at ip %s", amp_num, ip_addr)
+        for zone_num in range(1, 7):
+            zone_id = (amp_num * 10) + zone_num
+            entities.append(Ws66iZone(ws66i, sources, config_entry.entry_id, zone_id))
```

Comment:
Move this into its own function and wrap it in `hass.async_add_executor_job` so you aren't creating an executor job for each loop